pkg/storegateway/batch_series_test.go

```diff
@@ -265,6 +301,122 @@ func TestMergedBatchSet(t *testing.T) {
+					require.Equal(t, batches[expectedIdx], preloading.At())
+					expectedIdx++
+				}
+
+				// Ensure all batches have been returned.
+				require.NoError(t, preloading.Err())
+				require.Equal(t, len(batches), expectedIdx)
+			})
+		}
+	})
+
+	t.Run("should stop iterating once an error is found", func(t *testing.T) {
+		for preloadSize := 1; preloadSize <= len(batches)+1; preloadSize++ {
+			t.Run(fmt.Sprintf("preload size: %d", preloadSize), func(t *testing.T) {
+				source := newSliceLoadedBatchSet(errors.New("mocked error"), batches...)
+				preloading := newPreloadingBatchSet(context.Background(), preloadSize, source)
+
+				// Ensure expected batches are returned in order.
+				expectedIdx := 0
+				for preloading.Next() {
+					require.NoError(t, preloading.Err())
+					require.Equal(t, batches[expectedIdx], preloading.At())
+					expectedIdx++
+				}
+
+				// Ensure an error is returned at the end.
+				require.Error(t, preloading.Err())
+				require.Equal(t, len(batches), expectedIdx)
+			})
+		}
+	})
+
+	t.Run("should not leak preloading goroutine if caller doesn't iterated until the end of batches but context is canceled", func(t *testing.T) {
+		ctx, cancelCtx := context.WithCancel(context.Background())
+
+		source := newSliceLoadedBatchSet(errors.New("mocked error"), batches...)
+		preloading := newPreloadingBatchSet(ctx, 1, source)
+
+		// Just call Next() once.
+		require.True(t, preloading.Next())
+		require.NoError(t, preloading.Err())
+		require.Equal(t, batches[0], preloading.At())
+
+		// Cancel the context. At this point we expect Next() to return false.
+		cancelCtx()
+		require.False(t, preloading.Next())
+		require.NoError(t, preloading.Err())
+	})
+
+	t.Run("should not leak preloading goroutine if caller doesn't call Next() until false but context is canceled", func(t *testing.T) {
+		ctx, cancelCtx := context.WithCancel(context.Background())
+
+		source := newSliceLoadedBatchSet(errors.New("mocked error"), batches...)
+		preloading := newPreloadingBatchSet(ctx, 1, source)
+
+		// Just call Next() once.
+		require.True(t, preloading.Next())
+		require.NoError(t, preloading.Err())
+		require.Equal(t, batches[0], preloading.At())
+
+		// Cancel the context. Do NOT call Next() after canceling the context.
+		cancelCtx()
+	})
+}
+
+func TestPreloadingBatchSet_Concurrency(t *testing.T) {
+	const (
+		numRuns     = 100
+		numBatches  = 100
+		preloadSize = 10
+	)
+
+	// Create some batches.
+	batches := make([]loadedBatch, 0, numBatches)
+	for i := 0; i < numBatches; i++ {
+		batches = append(batches, loadedBatch{})
+	}
+
+	// Run many times to increase the likelihood to find a race (if any).
+	for i := 0; i < numRuns; i++ {
+		source := newSliceLoadedBatchSet(errors.New("mocked error"), batches...)
+		preloading := newPreloadingBatchSet(context.Background(), preloadSize, source)
+
+		for preloading.Next() {
+			require.NoError(t, preloading.Err())
+			require.NotNil(t, preloading.At())
```

Comment:
`preloading.At()` will never be nil. Not sure if we are testing anything. Maybe if you have `batches` that are non-zero-valued, you can use `require.NotZero()` on each instead of `require.NotNil()`

Reply:
Right, fixed with a `require.NotZero()`. My goal is to exercise `At()` too, to make sure no race condition is there as well.